actionpack/test/dispatch/request_test.rb

```diff
@@ -1051,6 +1051,43 @@ class RequestParameters < BaseRequestTest
     assert_raises(ActionController::BadRequest) { request.parameters }
   end
 
+  test "POST parameters containing invalid UTF8 character" do
+    data = "foo=%81E"
+    request = stub_request(
+      "REQUEST_METHOD" => "POST",
+      "CONTENT_LENGTH" => data.length,
+      "CONTENT_TYPE" => "application/x-www-form-urlencoded; charset=utf-8",
+      "rack.input" => StringIO.new(data)
+    )
+
+    err = assert_raises(ActionController::BadRequest) { request.parameters }
+
+    assert_predicate err.message, :valid_encoding?
+    assert_equal "Invalid request parameters: Invalid encoding for parameter: �E", err.message
+  end
+
+  test "query parameters specified as ASCII_8BIT encoded do not raise InvalidParameterError" do
+    request = stub_request("QUERY_STRING" => "foo=%81E")
+
+    ActionDispatch::Request::Utils.stub(:set_binary_encoding, { "foo" => "\x81E".b }) do
```

Comment:
We should not stub here since we are not testing the call of that method. We can assert the value of the parameters.

Reply:
I'm not sure I follow how to write this test without a stub. We want to test the case where the query parameters are specified as being binary encoded so we ensure we don't raise.

However, `ActionDispatch::Request::Utils.set_binary_encoding` won't actually encode our query params unless we have a controller (https://github.com/rails/rails/pull/40124/files#diff-a2247cc200a3f9cd3d0e5ef8d76c4df8R86) - it just returns them straight away, which causes the `Invalid encoding` error to get raised. So don't we want to stub `set_binary_encoding`?

Reply:
Ah, gotcha!

So we do already have test coverage for `Request::Utils.set_binary_encoding` via `actionpack/test/dispatch/request/multipart_params_parsing_test.rb` (see https://github.com/rails/rails/pull/40124/files#diff-897c5c6d482754e29f9571884ad3ce7fR10-R13 and https://github.com/rails/rails/pull/40124/files#diff-897c5c6d482754e29f9571884ad3ce7fL120-L124), as well as several tests in `actionpack/test/dispatch/routing_test.rb` (see https://github.com/rails/rails/blob/master/actionpack/test/dispatch/routing_test.rb#L4560-L4612), which act as full integration tests.

I can convert these request tests to also act as integration tests by adding controllers, but thought that was maybe not ideal since we have test coverage for the full flow already, and it seems reasonable to stub the result of `set_binary_encoding` for the request tests, relying on these tests just to assert that `GET` and `POST` rescue and re-raise the exception appropriately.